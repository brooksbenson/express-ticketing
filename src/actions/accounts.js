import db from '../firebase/firebase';

export const addAccount = ({ key, ...account }) => ({
  type: 'ADD_ACCOUNT',
  key,
  account
});

export const startAddAccount = account => {
  return async dispatch => {
    const { key } = await db.ref('account_data').push(account);
    dispatch(addAccount({ key, ...account }));
    return key;
  };
};

export const setAccounts = accounts => ({
  type: 'SET_ACCOUNTS',
  accounts
});

export const startSetAccounts = () => {
  return async dispatch => {
    const snap = await db.ref('account_data').once('value');
    dispatch(setAccounts(snap.val()));
  };
};

export const updateAccount = ({ key, ...update }) => ({
  type: 'UPDATE_ACCOUNT',
  key,
  update
});

export const startUpdateAccount = ({ key, ...update }) => {
  return async dispatch => {
    await db.ref(`account_data/${key}`).set(update);
    dispatch(updateAccount({ key, ...update }));
  };
};
