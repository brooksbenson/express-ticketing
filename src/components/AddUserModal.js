import React from 'react';
import Modal from 'react-modal';
import SearchBar from './SearchBar';
import userSelector from '../selectors/users';

Modal.setAppElement('body');

export default props => (
  <Modal
    className="add-user-modal"
    isOpen={props.isOpen}
    contentLabel="Add user modal"
    onRequestClose={props.onRequestClose}
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    closeTimeoutMS={200}
  >
    <h3> Add User </h3>
    <button onClick={props.onRequestClose}>
      <div />
      <div />
    </button>
    <SearchBar
      displayResults={!!props.searchString}
      onPick={props.onUserPick}
      onSearchChange={e => props.onSearchChange(e.target.value)}
      results={userSelector(props.users, props.searchString)}
      searchString={props.searchString}
    />
    <button onClick={props.onAddUser}>Add User</button>
  </Modal>
);
