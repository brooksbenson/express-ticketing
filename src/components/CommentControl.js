import React from 'react';

export default props => (
  <section className={props.className}>
    <h3> Comments </h3>
    <textarea
      onChange={e => props.onCommentChange(e.target.value)}
      value={props.comment}
    />
    <button onClick={props.onCommentSave}> Save Comment </button>
    <ul>
      {props.comments.map(c => (
        <li key={c.key}>
          <div>
            <span>{c.name}</span>
            <span>{c.date}</span>
          </div>
          <p>{c.text}</p>
        </li>
      ))}
    </ul>
  </section>
);
