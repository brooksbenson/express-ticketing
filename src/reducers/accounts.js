export default (state = [], action) => {
  switch (action.type) {
    case 'ADD_ACCOUNT':
      return [...state, action.account];
    case 'SET_ACCOUNTS':
      return action.accounts;
    default:
      return state;
  }
};
