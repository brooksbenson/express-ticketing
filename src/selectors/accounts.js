export default (accounts, search) => {
  const regex = new RegExp(search.split('').join('.*?'), 'i');
  return accounts.filter(a => regex.test(a.name));
};
