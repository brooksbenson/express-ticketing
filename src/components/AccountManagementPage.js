import React from 'react';
import { connect } from 'react-redux';
import { startAddAccount, startUpdateAccount } from '../actions/accounts';
import accountSelector from '../selectors/accounts';

export class AccountManagementPage extends React.Component {
  state = {
    accountKey: '',
    name: '',
    search: '',
    website: ''
  };

  onAccountClick = account => {
    this.setState(() => ({
      ...account,
      accountKey: account.key
    }));
  };

  onCancel = () => {
    this.setState(() => ({
      accountKey: '',
      name: '',
      website: ''
    }));
  };

  onNameChange = e => {
    const change = e.target.value;
    this.setState(() => ({
      name: change
    }));
  };

  onWebsiteChange = e => {
    const change = e.target.value;
    this.setState(() => ({
      website: change
    }));
  };

  onSave = e => {
    e.preventDefault();
    const defaults = { accountKey: '', name: '', website: '' };
    if (this.state.accountKey) {
      this.props.startUpdateAccount({ ...this.state }).then(() => {
        this.setState(() => ({ ...defaults }));
      });
    } else {
      const { accountKey, ...accountData } = this.state;
      this.props.startAddAccount({ ...accountData }).then(() => {
        this.setState(() => ({ ...defaults }));
      });
    }
  };

  onSearch = e => {
    const search = e.target.value;
    this.setState(() => ({ search }));
  };

  render() {
    const { accountKey, name, search, website } = this.state;
    return (
      <section className="content-container">
        <div className="manage-accounts content-innards">
          <h2 className="heading"> Manage Accounts </h2>
          <div className="manage-accounts__block">
            <h3 className="heading heading--secondary">
              {accountKey ? 'Update' : 'New'} Account
            </h3>
            <form className="manage-accounts__form" onSubmit={this.onSave}>
              <div className="manage-accounts__form-row">
                <span> Name </span>
                <input
                  className="input"
                  onChange={this.onNameChange}
                  type="text"
                  value={name}
                />
              </div>
              <div className="manage-accounts__form-row">
                <span> Website </span>
                <input
                  className="input"
                  onChange={this.onWebsiteChange}
                  type="text"
                  value={website}
                />
              </div>
              <button className="btn btn--secondary">Save</button>
              {accountKey && (
                <button className="btn btn-tertiary" onClick={this.onCancel}>
                  Cancel
                </button>
              )}
            </form>
          </div>
          <div className="manage-accounts__block">
            <h3 className="heading heading--secondary"> Accounts </h3>
            <input
              className="search"
              onChange={this.onSearch}
              placeholder="Search..."
              type="text"
              value={search}
            />
            <ul className="manage-accounts__list">
              {accountSelector(this.props.accounts, search).map(a => (
                <li
                  key={a.key}
                  onClick={() => {
                    this.onAccountClick(a);
                  }}
                >
                  <button>{a.name}</button>
                </li>
              ))}
            </ul>
          </div>
        </div>
      </section>
    );
  }
}

const mapStateToProps = ({ accounts }) => ({ accounts });
const mapDispatchToProps = dispatch => ({
  startAddAccount: async account => {
    await dispatch(startAddAccount(account));
  },
  startUpdateAccount: async update => {
    await dispatch(startUpdateAccount(update));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(
  AccountManagementPage
);
