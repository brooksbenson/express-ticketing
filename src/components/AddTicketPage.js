import React from 'react';
import { connect } from 'react-redux';
import SearchBar from './SearchBar';
import ContactController from './ContactController';
import { startAddTicket } from '../actions/tickets';
import {
  accountPick,
  accountSearchChange,
  contactCtrlDataChange,
  contactPick,
  contactSearchChange,
  descriptionChange,
  startNewContact,
  startUpdateContact,
  titleChange,
  toggleContactCtrl,
  urgencyChange
} from '../actions/new-ticket';

export const AddTicketPage = props => {
  const onUrgencyChange = e => {
    const change = e.target.value;
    if (/^(low|medium|high)$/i.test(change)) {
      props.urgencyChange(change);
    }
  };

  const onTitleChange = e => {
    const change = e.target.value;
    if (change.length <= 40) {
      props.titleChange(change);
    }
  };

  const submit = e => {
    e.preventDefault();
    props.startAddTicket({
      accountName: props.account.name,
      contactName: props.contact.name,
      contactEmail: props.contact.email,
      contactNumber: props.contact.number,
      assignments: [{ uid: props.user.uid, completed: false }],
      comments: [{ uid: props.user.uid, text: props.description }],
      date: Date.now(),
      status: 'Open',
      title: props.title,
      urgency: props.urgency
    });
  };

  return (
    <div className="ticket-form">
      <h2 className="heading"> Create Ticket </h2>
      <section className="ticket-form__block">
        <h3 className="heading-secondary"> Account </h3>
        <SearchBar
          className="ticket-form__search-bar"
          disabled={false}
          displayKey="name"
          onPick={props.accountPick}
          onSearchChange={props.accountSearchChange}
          placeholder="Search accounts..."
          searchString={props.accountSearchString}
          results={props.accountSearchResults}
          uniqueKey={'id'}
          values={props.accounts}
        />
        <div className="ticket-form__block-row">
          <SearchBar
            className="ticket-form__search-bar"
            disabled={props.account == null}
            displayKey="name"
            onPick={props.contactPick}
            onSearchChange={props.contactSearchChange}
            placeholder="Search contacts..."
            searchString={props.contactSearchString}
            results={props.contactSearchResults}
            uniqueKey={'email'}
            values={props.account ? props.account.contacts : []}
          />
          <ContactController {...props} />
        </div>
      </section>
      <section className="ticket-form__block">
        <h3 className="heading-secondary">Triage</h3>
        <div className="ticket-form__block-row">
          <input
            className="input"
            name="title"
            onChange={onTitleChange}
            placeholder="Title"
            type="text"
          />
          <select
            className="select"
            onChange={onUrgencyChange}
            value={props.urgency || 'urgency'}
          >
            <option value="urgency" disabled>
              Urgency
            </option>
            <option value="low">Low</option>
            <option value="medium">Medium</option>
            <option value="high">High</option>
          </select>
        </div>
        <textarea
          className="textarea ticket-form__textarea"
          name="description"
          onChange={e => props.descriptionChange(e.target.value)}
          placeholder="Description (optional)"
          value={props.description}
        />
      </section>
      <button className="btn btn--primary" onClick={submit}>
        Submit Ticket
      </button>
    </div>
  );
};

const mapStateToProps = ({ newTicket, user }) => ({ ...newTicket, user });
const mapDispatchToProps = dispatch => ({
  accountPick: pick => dispatch(accountPick(pick)),
  accountSearchChange: change => dispatch(accountSearchChange(change)),
  addTicket: ticket => dispatch(addTicket(ticket)),
  contactCtrlDataChange: change => dispatch(contactCtrlDataChange(change)),
  contactPick: pick => dispatch(contactPick(pick)),
  contactSearchChange: change => dispatch(contactSearchChange(change)),
  descriptionChange: change => dispatch(descriptionChange(change)),
  startNewContact: contact => dispatch(startNewContact(contact)),
  startUpdateContact: contact => dispatch(startUpdateContact(contact)),
  titleChange: change => dispatch(titleChange(change)),
  toggleContactCtrl: () => dispatch(toggleContactCtrl()),
  urgencyChange: change => dispatch(urgencyChange(change))
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTicketPage);
