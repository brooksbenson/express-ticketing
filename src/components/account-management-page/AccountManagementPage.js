import React from 'react';
import { connect } from 'react-redux';
import { startAddAccount } from '../../actions/accounts';

export class AccountManagementPage extends React.Component {
  state = {
    name: '',
    website: ''
  };

  onNameChange = e => {
    const change = e.target.value;
    this.setState(() => ({
      name: change
    }));
  };

  onWebsiteChange = e => {
    const change = e.target.value;
    this.setState(() => ({
      website: change
    }));
  };

  onSave = e => {
    e.preventDefault();
    this.props.startAddAccount({ ...this.state }).then(() => {
      this.setState(() => ({
        name: '',
        website: ''
      }));
    });
  };

  render() {
    return (
      <section className="content-container">
        <div className="manage-accounts content-innards">
          <h2 className="heading"> Manage Accounts </h2>
          <div className="manage-accounts__block">
            <h3 className="heading heading--secondary"> New Account </h3>
            <form className="manage-accounts__form">
              <div className="manage-accounts__form-row">
                <span> Name </span>
                <input
                  className="input"
                  onChange={this.onNameChange}
                  type="text"
                  value={this.state.name}
                />
              </div>
              <div className="manage-accounts__form-row">
                <span> Website </span>
                <input
                  className="input"
                  onChange={this.onWebsiteChange}
                  type="text"
                  value={this.state.website}
                />
              </div>
              <button className="btn btn--secondary" onClick={this.onSave}>
                Save
              </button>
            </form>
          </div>
          <div className="manage-accounts__block">
            <h3 className="heading heading--secondary"> Accounts </h3>
            <ul className="manage-accounts__list">
              {this.props.accounts.map(a => <li key={a.key}> {a.name} </li>)}
            </ul>
          </div>
        </div>
      </section>
    );
  }
}

const mapStateToProps = ({ accounts }) => ({ accounts });
const mapDispatchToProps = dispatch => ({
  startAddAccount: async account => {
    await dispatch(startAddAccount(account));
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(
  AccountManagementPage
);
