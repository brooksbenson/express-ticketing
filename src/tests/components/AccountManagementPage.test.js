import React from 'react';
import { shallow } from 'enzyme';
import { AccountManagementPage } from '../../components/AccountManagementPage';
import accounts from '../fixtures/accounts';

let wrapper;
let startAddAccountArg;
let startUpdateAccountArg;
beforeEach(() => {
  wrapper = shallow(
    <AccountManagementPage
      accounts={accounts}
      startAddAccount={account => {
        startAddAccountArg = account;
        return Promise.resolve();
      }}
      startUpdateAccount={account => {
        startUpdateAccountArg = account;
        return Promise.resolve();
      }}
    />
  );
});

test('AccountManagementPage should match snapshot', () => {
  expect(wrapper).toMatchSnapshot();
});

test('AccountManagementPage should setState on name change', () => {
  const value = 'Flower Motors';
  wrapper
    .find('input')
    .at(0)
    .simulate('change', { target: { value } });
  expect(wrapper.state('name')).toBe(value);
});

test('AccountManagementPage should setState on website change', () => {
  const value = 'www.flowerpower.com';
  wrapper
    .find('input')
    .at(1)
    .simulate('change', { target: { value } });
  expect(wrapper.state('website')).toBe(value);
});

test('AccountManagementPage should invoke startAddAccount prop with account data', () => {
  const state = {
    accountKey: '',
    name: '',
    website: 'www.universal.com',
    search: ''
  };
  wrapper.setState(state);
  wrapper.find('form').simulate('submit', { preventDefault: () => {} });
  const { accountKey, ...accountData } = state;
  expect(startAddAccountArg).toEqual(accountData);
});

test('AccountManagementPage should invoke startUpdateAccount prop', () => {
  const state = {
    accountKey: 'fjdka-lj',
    name: 'Universal Systems',
    website: 'www.universal.io',
    search: ''
  };
  wrapper.setState(state);
  wrapper.find('form').simulate('submit', { preventDefault: () => {} });
  expect(startUpdateAccountArg).toEqual(state);
});

test('AccountManagementPage should respond to account click correctly', () => {
  const [account] = accounts;
  wrapper
    .find('.manage-accounts__list')
    .children()
    .at(0)
    .simulate('click');
  expect(wrapper.state('accountKey')).toBe(account.key);
  expect(
    wrapper
      .find('form')
      .children()
      .find('button').length
  ).toBe(2);
  expect(
    wrapper
      .find('h3')
      .at(0)
      .text()
  ).toBe('Update Account');
});

test('should respond to cancel update correctly', () => {
  wrapper.setState({ ...accounts[0], accountKey: accounts[0].key });
  wrapper
    .find('button')
    .at(1)
    .simulate('click');
  expect(wrapper.state('accountKey')).toBeFalsy();
  expect(
    wrapper
      .find('form')
      .children()
      .find('button').length
  ).toBe(1);
  expect(
    wrapper
      .find('h3')
      .at(0)
      .text()
  ).toBe('New Account');
});

test('should setState on search change', () => {
  const value = 'walmart';
  wrapper
    .find('.search')
    .at(0)
    .simulate('change', { target: { value } });
  expect(wrapper.state('search')).toBe(value);
});

test('should change list on search', () => {
  const initialListLength = wrapper.find('li').length;
  wrapper.setState({ search: 'flower' });
  const searchListLength = wrapper.find('li').length;
  expect(initialListLength).toBeGreaterThan(searchListLength);
});
