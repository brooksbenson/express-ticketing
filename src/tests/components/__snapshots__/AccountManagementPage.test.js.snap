// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountManagementPage should match snapshot 1`] = `
<section
  className="content-container"
>
  <div
    className="manage-accounts content-innards"
  >
    <h2
      className="heading"
    >
       Manage Accounts 
    </h2>
    <div
      className="manage-accounts__block"
    >
      <h3
        className="heading heading--secondary"
      >
        New
         Account
      </h3>
      <form
        className="manage-accounts__form"
        onSubmit={[Function]}
      >
        <div
          className="manage-accounts__form-row"
        >
          <span>
             Name 
          </span>
          <input
            className="input"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <div
          className="manage-accounts__form-row"
        >
          <span>
             Website 
          </span>
          <input
            className="input"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
        <button
          className="btn btn--secondary"
        >
          Save
        </button>
      </form>
    </div>
    <div
      className="manage-accounts__block"
    >
      <h3
        className="heading heading--secondary"
      >
         Accounts 
      </h3>
      <input
        className="search"
        onChange={[Function]}
        placeholder="Search..."
        type="text"
        value=""
      />
      <ul
        className="manage-accounts__list"
      >
        <li
          key="jfkda;"
          onClick={[Function]}
        >
          <button>
            Flower Motors
          </button>
        </li>
        <li
          key="kfjdka"
          onClick={[Function]}
        >
          <button>
            Prophecy Systems
          </button>
        </li>
        <li
          key="fjdkal"
          onClick={[Function]}
        >
          <button>
            Dynamic Foods
          </button>
        </li>
      </ul>
    </div>
  </div>
</section>
`;
